### ISSUE TITLE
Module conversion tracking issue / notes

### ISSUE THREAD

**Issue by jakebailey** [microsoft/TypeScript#49332]
I'm using this thread to dump my brain and to note problems as I solve them (or to discuss).

Issues of note:

-   #49037
-   #27891
-   #46744 (Eli's old notes)
-   #35210
-   #46567

### Changes since the previous effort

-   The "typeformer" (the code that does the transform) now lives at https://github.com/jakebailey/typeformer, and has been rewritten using [ts-morph](https://github.com/dsherret/ts-morph) to preserve comments and formatting.
    -   The old typeformer use TS's own transform and emitter, but the emitter is not careful enough to preform TS-to-TS transformations (I have a list of bugs I've discovered, which don't matter for JS emit)
-   My WIP changes are semi-automatically pushed to a stack of PRs on my fork, starting [here](https://github.com/jakebailey/TypeScript/pull/1/files).
    -   It's "best" to look at the difference between the "indent" commit and the last commit, e.g. [this diff](https://github.com/jakebailey/TypeScript/compare/transform-stack-commit-2...transform-stack-commit-10) (but note that this link may not be up to date, is is as of now).
    -   Please don't comment on these PRs; the script I run to create this stack is not very smart and will force-push if I insert a new commit in the middle. I wish I could just use `gerrit`, but...
-   I have removed the esbuild step from the previous effort; not because we don't plan to use it, but because it's simplest to start by using raw modules and then tweaking the output later. If we're going to be modules, I'd hope we can be run through a bundler successfully.
-   The faked namespace modules (barrel modules that at runtime behave very similarly to the old namespaces) have been moved to a dedicated folder in each project, rather than being dumped at the top level.
-   #46944 has to be reverted; this code doesn't seem to compile properly when actually built.

### typeformer bugs

-   The typeformer's `inlineImports` step converts `globalThis.assert` into `assert`, because it thinks it's a global function that should be settable. Technically true, but if we wanted that, we probably wouldn't have written `globalThis` explicitly. This is just a bug when I rewrote the step.
-   Sometimes `#region` comments are dropped. Likely a bug that can be worked around like other ts-morph oddities to do with comment preservation (which it is almost always good at, but sometimes not).
-   Sometimes comments get duplicated, e.g. `// DestructuringAssignmentType` in `src/compiler/types.ts`.
-   The output path is determined by the project name; this was changed in my iteration of the typeformer, but should be hand-tweaked back to the right path, e.g. the debug code (`dbg.ts`) that is dynamically loaded in, or `loggedIO`.

### Unsolved problems

-   My version of the typeformer's "inlineImports" pass is very aggressive, importing everything from their original declaration if possible. This is closest to how we would write the code by hand (if we had done so from the start), but the circularity of the codebase has broken things.
    -   e.g., `debug.ts` imports `semver.ts`, but the `Semver` class creates a static property `zero`, which calls the `Semver` constructor, which then contains `Debug.assert` calls.
    -   This only happened because the top of `debug.ts` contains `import { Semver } from './semver`, whereas before, the code didn't do this but instead referred to the namespace, so never actually caused that file to be evaluated.
    -   It's totally possible that we'd hit this circularity issue today, except that most code is not at the top level and avoids the problem entirely, or has worked "magically" thanks to the order we happen to merge namespaces in.
    -   We could just import these from the namespaces (like #46567 does), which I think would fix this, but this is really unnatural code structure and auto-imports will not function properly.
-   The `System` interface (aka `ts.sys`) is set (`ts.sys = ...`) in order to support the browser, but in the module version, modules do `import { sys } from '...'`), which makes it very difficult to do like this. What can do do here?
-   Additionally, the `System` interface has `getExecutingFilePath` (among other functions), which only makes sense for TS packaged as a self contained file.
-

### TODOs

-   [ ] Figure out what to do about the API; the past effort talks about API extractor, which is the right thing to do, but we need to make sure that we produce an output that describes the API as namespaces, not as reexported modules. #4529 makes this hard.
-   [ ] We need to create the entrypoints for the new exported API; right now it just outputs as `outDir`, but I don't think we intend to just allow everyone to import our internals directly. `outFile` made a file that did the right thing before, and after we just need to write something like `tsc.js` which exports `*` from `_namespaces/ts.ts` or something.
    -   See also `exportAsModule`, which hacks together a `module.exports = ...` line rather than emitting it via `export`.
-   [ ] Deprecations were previously done via namespace merging (i.e. deprecating "Map" in favor of "ESMap"); these are very messy when not coming up with a single file.
-   [ ] `registerRefactor` uses side effects of file loading to register refactorings. This technically works OK after modification if the fake namespace modules still import these files, but it'd be better to refactor this to work like the transforms (which are just imported in other files instead).
-   [ ] Loads of type names are duplicated; this previously didn't matter because you'd just fully qualify names for another namespace, but now that things are directly imported, you end up with these random fully-qualified names which would be better done differently. For example, `fakeHosts.ts` implements certain interfaces, but does so with classes named the same as the interface. Maybe just do `FakeSomeInterface`?

### Other stuff

-   We should be able to hide the big refactor when it's merged using a `.git-blame-ignore` file. See:
    -   https://github.com/microsoft/vscode/blob/main/.git-blame-ignore
    -   https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view

**Comment by weswigham** [comment]
> My version of the typeformer's "inlineImports" pass is very aggressive, importing everything from their original declaration if possible. This is closest to how we would write the code by hand (if we had done so from the start), but the circularity of the codebase has broken things.
e.g., debug.ts imports semver.ts, but the Semver class creates a static property zero, which calls the Semver constructor, which then contains Debug.assert calls.
This only happened because the top of debug.ts contains import { Semver } from './semver, whereas before, the code didn't do this but instead referred to the namespace, so never actually caused that file to be evaluated.
It's totally possible that we'd hit this circularity issue today, except that most code is not at the top level and avoids the problem entirely, or has worked "magically" thanks to the order we happen to merge namespaces in.
We could just import these from the namespaces (like https://github.com/microsoft/TypeScript/pull/46567 does), which I think would fix this, but this is really unnatural code structure and auto-imports will not function properly.

I'd fixed this by hand a few times in my version - I usually swap deep imports for barrel file imports and reorder the barrel imports to match our tsconfig order, since that represents the canonical execution order we desire. Sometimes some barrel imports need to be replaced with deep imports to break circularities, but it wasn't too many, iirc.

This:
> e.g., debug.ts imports semver.ts, but the Semver class creates a static property zero, which calls the Semver constructor, which then contains Debug.assert calls.

? I don't see a `Semver` reference in `debug.ts` locally, and `semver` comes after `debug` in our tsconfig, so `semver`'s never available during `debug`'s setup (not that it seems to need to be?).

> or has worked "magically" thanks to the order we happen to merge namespaces in.

Nothing _too_ magic about it. The `tsconfig` order is curated to be a correct execution order - so long as barrel files reproduce the same order (either automatically or by hand), we should be equivalent, at least.

> The System interface (aka ts.sys) is set (ts.sys = ...) in order to support the browser, but in the module version, modules do import { sys } from '...'), which makes it very difficult to do like this. What can do do here?

We have a `setSys` function in `sys` already as part of my work years ago - we should swap external stuff over to using that if it isn't already.

> Additionally, the System interface has getExecutingFilePath (among other functions), which only makes sense for TS packaged as a self contained file.

It still makes sense for a tree of files, imo - intuitively I feel like it should be the entrypoint file, or whatever was initially executed by `node`. (be it tsc.js or tsserver.js or bin/whatever)

**Comment by jakebailey** [comment]
> ? I don't see a Semver reference in debug.ts locally, and semver comes after debug in our tsconfig, so semver's never available during debug's setup (not that it seems to need to be?).

I meant to say `Version`, which is defined in `semver.ts`. It's referenced many times. (I was confusing it with the `Semver` package everyone else uses.)

> I'd fixed this by hand a few times in my version - I usually swap deep imports for barrel file imports and reorder the barrel imports to match our tsconfig order, since that represents the canonical execution order we desire. Sometimes some barrel imports need to be replaced with deep imports to break circularities, but it wasn't too many, iirc.

> Nothing too magic about it. The tsconfig order is curated to be a correct execution order - so long as barrel files reproduce the same order (either automatically or by hand), we should be equivalent, at least.

My overall goal was to end up with a codebase that felt like one that had been written with modules; this means eliminating these barrel imports from files. Otherwise, auto-imports will not function properly because auto-imports will prefer importing directly from the declaring modules. It also means not requiring that a given file import each file in a specific order either, given everyone (including our own import organizer) will not preserve this order.

> Nothing too magic about it. The tsconfig order is curated to be a correct execution order - so long as barrel files reproduce the same order (either automatically or by hand), we should be equivalent, at least.

This is going to be rough to maintain in the long term, because I don't think the average person knows that this ordering has any impact. I sure didn't; I've never used a project that listed its files explicitly, only globs. I would hope that we can stop having so few files just to not have to add more, and instead break the code apart a bit. Especially if perf improves because we're not running every file in a closure where each access to the old namespace was actually fully-qualified under the hood.

> We have a setSys function in sys already as part of my work years ago - we should swap external stuff over to using that if it isn't already.

My understanding had been that if one has a `import { sys } from "..."` that you're stuck with whatever value existed at import time, but searching it now, I guess `export let ...` is some sort of magical "live" view, which does fix this. That's good.

> It still makes sense for a tree of files, imo - intuitively I feel like it should be the entrypoint file, or whatever was initially executed by node. (be it tsc.js or tsserver.js or bin/whatever)

Yep, this is more of a plumbing problem. I just wanted to note it so I could delete my random TODOs.

**Comment by weswigham** [comment]
> This is going to be rough to maintain in the long term

I mean, it already is, it's one of the hidden costs if using namespaces. It just only comes up when someone finds they want to adjust the order (because, eg, they want to start using `utilities` somewhere new or something). Likewise, it'll only come up with modules when someone wants to import from a new module that's already in use elsewhere, which, in practice, is pretty rare.

**Comment by weswigham** [comment]
> this means eliminating these barrel imports from files

So this is a style argument, really. Imo, you're _supposed_ to import from the barrel files for everything other than your immediately adjacent files (and we've talked about our autoimport behavior re: this before, and the only way we'll get better at it is honestly using it ourselves and seeing the upsides and downsides of each approach). Still, our first concern should probably be retaining a functional execution order (T.T) and not a specific style - that can come later.

**Comment by jakebailey** [comment]
My next step is to revert my changes that made imports really aggressive in favor of importing from the barrels, just to confirm that things run again (and to double check the state of auto-imports), so that should hopefully get me back to a working codebase.

I was really hopeful that we could write things like everyone else, but it's a shame about the execution ordering.

**Comment by jakebailey** [comment]
## 2022-08-29

I took a few months off of this to work on other issues, but have recently picked this back up again with the intent to make this breaking packaging change for 5.0 (it seems like a good idea to not do this on a minor release).

Last week, I hit a major milestone and got the test suite to run; all tests pass except the API tests (which don't pass because I haven't started on fixing the d.ts files yet).

- The entire build is just `generateLibs` and `tsc -b ./src`, which is a great simplification from our current state. I expect this to only expand for things like `api-extractor` and potentially a bundler if we still need to support a single-file for some outputs. The build time is roughly the same as `main` for a clean build of every project, but a good deal faster than `main` when making an incremental change to `checker.ts`.
- Getting tests running required a reordering of things in the namespaces, as well as moving some helper code around such that tests do not import other tests.
- The tests only work via the serial test runner, i.e. what's built into mocha. Our own custom parallel runner does not work. I am investigating this, however, mocha has a built-in parallel test runner. Our own was built before it didn't, so I'm going to see if we can drop our own special runner and just stick to upstream.
- I had to bring the module transform back to importing things from the faked namespaces (as noted in my previous comment); this is not my preference, but our code is currently too circular / execution order dependent to use direct imports.
  - It's likely that we can just start with these namespace imports first, then fix the circularities over time. There's a tool that the VS Code team used to analyze their codebase for these sorts of issues, and would likely work on our own once it uses imports.
- `getExecutingFilePath` is still replaced with a dummy implementation; reasonably I expect this API to change to something more like `getTypeScriptLibRoot` or similar such that the path to our lib folder can be provided rather than being deduced relative to `tsc.js` or something.

**Comment by jakebailey** [comment]
I posted the above early in the day; I did get the parallel test suite working. Mocha's parallel runner unfortunately can't be used with our repo in its current state as it expects individual test files, which it then runs in parallel (never any actual tests within a file in parallel).

We do our own thing and end up with effectively one file. Potentially, we can switch to something "more normal", somehow, but it's no longer a blocker. If we were using something like jest or ava, they could potentially run tests within a file in parallel, but that's a much larger change.

**Comment by jakebailey** [comment]
## 2022-08-31

I have the parallel test runner working; it came down to the fact that the entrypoint I was using for the test runner was the namespace file, which broke because `runner.ts` exported a mutable variable to another file, but `runner.ts` hadn't yet "returned" to the namespace object's `require('./runner.ts')` to add the binding to its own exports, so it appeared to be undefined. This wouldn't happen if we used direct imports, but making the entrypoint be `runner.js` was enough.

While working on this, I prototyped a Jest runner that could run our compiler tests instead of the relatively large bit of code we wrote to make Mocha work; I think that's a good idea for the future. I only tried it because I was frustrated trying to get our own parallel runner to work (so was looking for any alternative), but I'll defer that to another day now that I have all tests running besides the API tests.

Next up is api-extractor, and getting some proper entrypoints for things like `tsc.js`, `tsserver.js`, `typescript.js`, etc, which we should now be able to build unconditionally in build mode as they will be small.

**Comment by jakebailey** [comment]
## 2022-09-02

I have the entrypoints set up and working for the libraries/executables, and a basic api-extractor config for one of them. Unfortunately, things seem to fail on that front. I've reported https://github.com/microsoft/rushstack/issues/3612.

We'll have to eventually call api-extractor programatically; one transformation we will need to do again is to convert our `const enum` to `enum`; I doubt api-extractor does this for us.

**Comment by jakebailey** [comment]
## 2022-09-12

The aforementioned `import()` thing has to do with #44044 and #49730; the cases that crash api-extractor are ones where we don't emit an import statement because the original file didn't have one, so emit an import expression instead. api-extractor shouldn't be crashing for these cases, but I'm going to try and explicitly annotate the offending pieces of code and see how far I get.

**Comment by jakebailey** [comment]
## 2022-09-14

I've gotten things far along enough to run perf benchmarks, and I'm overwealmingly happy to share the results.

- [Emitting CJS, 1:1 output to source](https://github.com/microsoft/TypeScript/pull/50765#issuecomment-1247225287), we see a performance loss of anywhere from 15 to 55 percent. 
- [Emitting ESM, 1:1 output to source](https://github.com/microsoft/TypeScript/pull/50765#issuecomment-1247365157), we see a performance _gain_ of 3-9%. 🎉

This difference is very likely due to the fact that we are going through our export helpers, which are like our old namespace accesses, but extra slow. Compare that to ESM, which is faster as it doesn't use those import helpers, but since we import things like helper functions (`isIdentifier`, `createParser`, etc), we see a performance boost as we are no longer accessing via an object for each invocation, instead using a local binding created by the import statement.

Now, for esbuild:

- [Running esbuild on our CJS output](https://github.com/microsoft/TypeScript/pull/50765#issuecomment-1246285643), we see about the same performance loss as individual CJS files (maybe slightly worse). Not surprising, our import helpers are still present, on top of esbuild's.
- [Running esbuild directly on our source files](https://github.com/microsoft/TypeScript/pull/50765#issuecomment-1246231797) (no tsc emit), we see an **incredible 7-25% performance boost!** Looking at the output file, esbuild was able to lift all modules to the top level, meaning all accesses to functions in other modules are effectively local; no overhead. What's even more incredible is that this is without lowering const/let to var, something that esbuild currently does not support. (Could we go even faster?)
- [Running esbuild on our ESM output](https://github.com/microsoft/TypeScript/pull/50765#issuecomment-1247409184), we see about a 6-13% performance boost. This is somewhat surprising, as this is a case where we were using var instead of let/const, so I would have expected it to be even faster than the previous case.

Previously, we had intended to try and ship as individual files, a breaking change for 5.0. Since we're not likely to switch away from CJS, the performance penalty there is pretty unfortunate. But, if we just bundle our output, we would be producing a package that's still compatible with the previous one, and turns out to be 7-25% faster, _and_ has an extremely fast build that we can make use of during development.

--

At this point, most things are working; I have some remaining tasks:

- api-extractor; the bug I'm hitting needs to be minimized and reported. Declarations that are marked as `@internal` aren't stripped from the namespace objects produced for our namespace-barrel modules. Likely an oversight, as `export *` support is new.
- Random loose end endpoints need to be handled, like the typings installer, cancellation token, etc.
- `dynamicImportCompat` needs to be checked (and maybe rethought).
- `buildProtocol` needs to be rethought (do we need to keep providing protocol.d.ts?)
- `createPlaygroundBuild` needs to be checked to make sure it works; this was tricker when I was planning to output individual modules, but probably no longer a problem if we are bundling.
- The build needs to be massively cleaned up. Things are a lot simpler, but I haven't been deleting anything from the gulpfile quite yet.
- Whatever else comes up.

**Comment by jakebailey** [comment]
## 2022-09-19

On #50811, I tested what it'd be like to downlevel `let`/`const` to `var` on esbuild's output. Since it doesn't support that itself, I just had our build run `@babel/plugin-transform-block-scoping` on the output.

The additional performance boost was as expected, roughly matching the performance difference I noted when trying to switch our target up past ES5 on #50022, or about 3%. Most of that difference comes from improving parsing performance by 5-9%, which indicates that it's doing a lot in the TDZ. The other parts of the compiler don't improve as greatly, but do generally get a few percent faster.

Running babel on the output of esbuild works, but is definitely a lot slower than esbuild itself. I don't think we need to downlevel everything to ES5, but `--supported=let-and-const:false` support in esbuild would be very helpful.

<details><summary> Benchmark results, bundled before and after let/const downleveling:</summary><p>
<table border="0" cellpadding="0" cellspacing="0">
<thead>
<tr>
<th align="left">Metric</th>
<th align="right">50811</th>
<th align=right>50811</th>
<th align="right">Delta</th>
<th align="right">Best</th>
<th align="right">Worst</th>
</tr>
</thead>
<tr><th align="left" colspan="6">Angular - node (v14.15.1, x64)</th></tr>
<tbody><tr>
<td align="left" valign="middle"><nobr>Memory used</nobr></td>
<td align="right" valign="top">328,472k<br /><sup>±0.01%</sup></td>
<td align="right" valign="top">328,287k<br /><sup>±0.00%</sup></td>
<td align="right" valign="top">-185k<br /><sup>-0.06%</sup></td>
<td align="right" valign="top">328,244k</td>
<td align="right" valign="top">328,312k</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Parse Time</nobr></td>
<td align="right" valign="top">2.06s<br /><sup>±0.52%</sup></td>
<td align="right" valign="top">1.93s<br /><sup>±0.41%</sup></td>
<td align="right" valign="top">-0.13s<br /><sup>-6.36%</sup></td>
<td align="right" valign="top">1.91s</td>
<td align="right" valign="top">1.95s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Bind Time</nobr></td>
<td align="right" valign="top">0.70s<br /><sup>±0.57%</sup></td>
<td align="right" valign="top">0.69s<br /><sup>±0.71%</sup></td>
<td align="right" valign="top">-0.01s<br /><sup>-0.72%</sup></td>
<td align="right" valign="top">0.69s</td>
<td align="right" valign="top">0.71s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Check Time</nobr></td>
<td align="right" valign="top">5.35s<br /><sup>±0.39%</sup></td>
<td align="right" valign="top">5.28s<br /><sup>±0.39%</sup></td>
<td align="right" valign="top">-0.07s<br /><sup>-1.31%</sup></td>
<td align="right" valign="top">5.25s</td>
<td align="right" valign="top">5.33s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Emit Time</nobr></td>
<td align="right" valign="top">5.20s<br /><sup>±0.82%</sup></td>
<td align="right" valign="top">5.15s<br /><sup>±0.56%</sup></td>
<td align="right" valign="top">-0.06s<br /><sup>-1.08%</sup></td>
<td align="right" valign="top">5.07s</td>
<td align="right" valign="top">5.21s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Total Time</nobr></td>
<td align="right" valign="top">13.31s<br /><sup>±0.37%</sup></td>
<td align="right" valign="top">13.05s<br /><sup>±0.26%</sup></td>
<td align="right" valign="top">-0.25s<br /><sup>-1.92%</sup></td>
<td align="right" valign="top">12.98s</td>
<td align="right" valign="top">13.15s</td>
</tr>
</tbody><tr><th align="left" colspan="6">Compiler-Unions - node (v14.15.1, x64)</th></tr>
<tbody><tr>
<td align="left" valign="middle"><nobr>Memory used</nobr></td>
<td align="right" valign="top">182,299k<br /><sup>±0.44%</sup></td>
<td align="right" valign="top">182,374k<br /><sup>±0.39%</sup></td>
<td align="right" valign="top">+75k<br /><sup>+0.04%</sup></td>
<td align="right" valign="top">181,976k</td>
<td align="right" valign="top">185,246k</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Parse Time</nobr></td>
<td align="right" valign="top">0.89s<br /><sup>±0.65%</sup></td>
<td align="right" valign="top">0.81s<br /><sup>±0.37%</sup></td>
<td align="right" valign="top">-0.08s<br /><sup>-8.91%</sup></td>
<td align="right" valign="top">0.80s</td>
<td align="right" valign="top">0.81s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Bind Time</nobr></td>
<td align="right" valign="top">0.46s<br /><sup>±0.87%</sup></td>
<td align="right" valign="top">0.45s<br /><sup>±0.50%</sup></td>
<td align="right" valign="top">-0.01s<br /><sup>-2.18%</sup></td>
<td align="right" valign="top">0.44s</td>
<td align="right" valign="top">0.45s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Check Time</nobr></td>
<td align="right" valign="top">6.22s<br /><sup>±0.67%</sup></td>
<td align="right" valign="top">6.16s<br /><sup>±0.50%</sup></td>
<td align="right" valign="top">-0.07s<br /><sup>-1.06%</sup></td>
<td align="right" valign="top">6.10s</td>
<td align="right" valign="top">6.22s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Emit Time</nobr></td>
<td align="right" valign="top">1.99s<br /><sup>±0.98%</sup></td>
<td align="right" valign="top">1.96s<br /><sup>±0.76%</sup></td>
<td align="right" valign="top">-0.03s<br /><sup>-1.31%</sup></td>
<td align="right" valign="top">1.94s</td>
<td align="right" valign="top">2.01s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Total Time</nobr></td>
<td align="right" valign="top">9.56s<br /><sup>±0.60%</sup></td>
<td align="right" valign="top">9.38s<br /><sup>±0.32%</sup></td>
<td align="right" valign="top">-0.18s<br /><sup>-1.87%</sup></td>
<td align="right" valign="top">9.30s</td>
<td align="right" valign="top">9.43s</td>
</tr>
</tbody><tr><th align="left" colspan="6">Monaco - node (v14.15.1, x64)</th></tr>
<tbody><tr>
<td align="left" valign="middle"><nobr>Memory used</nobr></td>
<td align="right" valign="top">315,704k<br /><sup>±0.01%</sup></td>
<td align="right" valign="top">315,437k<br /><sup>±0.01%</sup></td>
<td align="right" valign="top">-267k<br /><sup>-0.08%</sup></td>
<td align="right" valign="top">315,401k</td>
<td align="right" valign="top">315,488k</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Parse Time</nobr></td>
<td align="right" valign="top">1.56s<br /><sup>±0.48%</sup></td>
<td align="right" valign="top">1.47s<br /><sup>±0.35%</sup></td>
<td align="right" valign="top">-0.09s<br /><sup>-5.52%</sup></td>
<td align="right" valign="top">1.46s</td>
<td align="right" valign="top">1.48s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Bind Time</nobr></td>
<td align="right" valign="top">0.63s<br /><sup>±0.94%</sup></td>
<td align="right" valign="top">0.62s<br /><sup>±0.55%</sup></td>
<td align="right" valign="top">-0.01s<br /><sup>-1.74%</sup></td>
<td align="right" valign="top">0.62s</td>
<td align="right" valign="top">0.63s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Check Time</nobr></td>
<td align="right" valign="top">5.14s<br /><sup>±0.44%</sup></td>
<td align="right" valign="top">5.12s<br /><sup>±0.50%</sup></td>
<td align="right" valign="top">-0.02s<br /><sup>-0.29%</sup></td>
<td align="right" valign="top">5.06s</td>
<td align="right" valign="top">5.18s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Emit Time</nobr></td>
<td align="right" valign="top">2.79s<br /><sup>±0.71%</sup></td>
<td align="right" valign="top">2.74s<br /><sup>±0.42%</sup></td>
<td align="right" valign="top">-0.05s<br /><sup>-1.72%</sup></td>
<td align="right" valign="top">2.71s</td>
<td align="right" valign="top">2.77s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Total Time</nobr></td>
<td align="right" valign="top">10.12s<br /><sup>±0.29%</sup></td>
<td align="right" valign="top">9.96s<br /><sup>±0.32%</sup></td>
<td align="right" valign="top">-0.16s<br /><sup>-1.60%</sup></td>
<td align="right" valign="top">9.89s</td>
<td align="right" valign="top">10.04s</td>
</tr>
</tbody><tr><th align="left" colspan="6">TFS - node (v14.15.1, x64)</th></tr>
<tbody><tr>
<td align="left" valign="middle"><nobr>Memory used</nobr></td>
<td align="right" valign="top">280,802k<br /><sup>±0.01%</sup></td>
<td align="right" valign="top">278,434k<br /><sup>±0.01%</sup></td>
<td align="right" valign="top">-2,367k<br /><sup>-0.84%</sup></td>
<td align="right" valign="top">278,374k</td>
<td align="right" valign="top">278,495k</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Parse Time</nobr></td>
<td align="right" valign="top">1.32s<br /><sup>±0.76%</sup></td>
<td align="right" valign="top">1.26s<br /><sup>±1.47%</sup></td>
<td align="right" valign="top">-0.06s<br /><sup>-4.91%</sup></td>
<td align="right" valign="top">1.23s</td>
<td align="right" valign="top">1.31s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Bind Time</nobr></td>
<td align="right" valign="top">0.58s<br /><sup>±0.63%</sup></td>
<td align="right" valign="top">0.59s<br /><sup>±0.62%</sup></td>
<td align="right" valign="top">+0.00s<br /><sup>+0.17%</sup></td>
<td align="right" valign="top">0.58s</td>
<td align="right" valign="top">0.59s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Check Time</nobr></td>
<td align="right" valign="top">5.03s<br /><sup>±0.43%</sup></td>
<td align="right" valign="top">4.97s<br /><sup>±0.46%</sup></td>
<td align="right" valign="top">-0.05s<br /><sup>-1.07%</sup></td>
<td align="right" valign="top">4.92s</td>
<td align="right" valign="top">5.02s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Emit Time</nobr></td>
<td align="right" valign="top">2.95s<br /><sup>±0.56%</sup></td>
<td align="right" valign="top">2.71s<br /><sup>±0.54%</sup></td>
<td align="right" valign="top">-0.24s<br /><sup>-8.07%</sup></td>
<td align="right" valign="top">2.69s</td>
<td align="right" valign="top">2.75s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Total Time</nobr></td>
<td align="right" valign="top">9.89s<br /><sup>±0.28%</sup></td>
<td align="right" valign="top">9.53s<br /><sup>±0.30%</sup></td>
<td align="right" valign="top">-0.36s<br /><sup>-3.60%</sup></td>
<td align="right" valign="top">9.49s</td>
<td align="right" valign="top">9.60s</td>
</tr>
</tbody><tr><th align="left" colspan="6">material-ui - node (v14.15.1, x64)</th></tr>
<tbody><tr>
<td align="left" valign="middle"><nobr>Memory used</nobr></td>
<td align="right" valign="top">427,373k<br /><sup>±0.00%</sup></td>
<td align="right" valign="top">427,047k<br /><sup>±0.00%</sup></td>
<td align="right" valign="top">-326k<br /><sup>-0.08%</sup></td>
<td align="right" valign="top">427,006k</td>
<td align="right" valign="top">427,094k</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Parse Time</nobr></td>
<td align="right" valign="top">1.87s<br /><sup>±0.46%</sup></td>
<td align="right" valign="top">1.77s<br /><sup>±0.42%</sup></td>
<td align="right" valign="top">-0.10s<br /><sup>-5.25%</sup></td>
<td align="right" valign="top">1.76s</td>
<td align="right" valign="top">1.79s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Bind Time</nobr></td>
<td align="right" valign="top">0.53s<br /><sup>±0.65%</sup></td>
<td align="right" valign="top">0.53s<br /><sup>±0.71%</sup></td>
<td align="right" valign="top">-0.00s<br /><sup>-0.38%</sup></td>
<td align="right" valign="top">0.52s</td>
<td align="right" valign="top">0.53s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Check Time</nobr></td>
<td align="right" valign="top">12.00s<br /><sup>±0.48%</sup></td>
<td align="right" valign="top">11.91s<br /><sup>±0.35%</sup></td>
<td align="right" valign="top">-0.09s<br /><sup>-0.74%</sup></td>
<td align="right" valign="top">11.81s</td>
<td align="right" valign="top">11.99s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Emit Time</nobr></td>
<td align="right" valign="top">0.00s<br /><sup>±0.00%</sup></td>
<td align="right" valign="top">0.00s<br /><sup>±0.00%</sup></td>
<td align="right" valign="top">0.00s<br /><sup> NaN%</sup></td>
<td align="right" valign="top">0.00s</td>
<td align="right" valign="top">0.00s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Total Time</nobr></td>
<td align="right" valign="top">14.40s<br /><sup>±0.40%</sup></td>
<td align="right" valign="top">14.21s<br /><sup>±0.32%</sup></td>
<td align="right" valign="top">-0.19s<br /><sup>-1.30%</sup></td>
<td align="right" valign="top">14.10s</td>
<td align="right" valign="top">14.31s</td>
</tr>
</tbody><tr><th align="left" colspan="6">xstate - node (v14.15.1, x64)</th></tr>
<tbody><tr>
<td align="left" valign="middle"><nobr>Memory used</nobr></td>
<td align="right" valign="top">517,236k<br /><sup>±0.01%</sup></td>
<td align="right" valign="top">516,843k<br /><sup>±0.00%</sup></td>
<td align="right" valign="top">-394k<br /><sup>-0.08%</sup></td>
<td align="right" valign="top">516,784k</td>
<td align="right" valign="top">516,900k</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Parse Time</nobr></td>
<td align="right" valign="top">2.68s<br /><sup>±0.57%</sup></td>
<td align="right" valign="top">2.53s<br /><sup>±0.67%</sup></td>
<td align="right" valign="top">-0.15s<br /><sup>-5.63%</sup></td>
<td align="right" valign="top">2.50s</td>
<td align="right" valign="top">2.57s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Bind Time</nobr></td>
<td align="right" valign="top">0.85s<br /><sup>±0.86%</sup></td>
<td align="right" valign="top">0.84s<br /><sup>±0.68%</sup></td>
<td align="right" valign="top">-0.01s<br /><sup>-0.94%</sup></td>
<td align="right" valign="top">0.83s</td>
<td align="right" valign="top">0.85s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Check Time</nobr></td>
<td align="right" valign="top">1.47s<br /><sup>±0.50%</sup></td>
<td align="right" valign="top">1.46s<br /><sup>±0.52%</sup></td>
<td align="right" valign="top">-0.00s<br /><sup>-0.27%</sup></td>
<td align="right" valign="top">1.45s</td>
<td align="right" valign="top">1.48s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Emit Time</nobr></td>
<td align="right" valign="top">0.07s<br /><sup>±4.13%</sup></td>
<td align="right" valign="top">0.07s<br /><sup>±3.23%</sup></td>
<td align="right" valign="top">-0.00s<br /><sup>-4.17%</sup></td>
<td align="right" valign="top">0.06s</td>
<td align="right" valign="top">0.07s</td>
</tr><tr>
<td align="left" valign="middle"><nobr>Total Time</nobr></td>
<td align="right" valign="top">5.08s<br /><sup>±0.37%</sup></td>
<td align="right" valign="top">4.91s<br /><sup>±0.51%</sup></td>
<td align="right" valign="top">-0.17s<br /><sup>-3.39%</sup></td>
<td align="right" valign="top">4.87s</td>
<td align="right" valign="top">4.98s</td>
</tr>
</tbody>
</table>  

**System**  
<table border="0" cellpadding="0" cellspacing="0">
<tbody>
<tr><th align="right">Machine Name</th><td>ts-ci-ubuntu</td></tr>
<tr><th align="right">Platform</th><td>linux 4.4.0-210-generic</td></tr>
<tr><th align="right">Architecture</th><td>x64</td></tr>
<tr><th align="right">Available Memory</th><td>16 GB</td></tr>
<tr><th align="right">CPUs</th><td>4 × Intel(R) Core(TM) i7-4770 CPU &#64; 3.40GHz</td></tr>
</tbody>
</table>  

**Hosts**  
* node (v14.15.1, x64)
  

**Scenarios**  
* Angular - node (v14.15.1, x64)
* Compiler-Unions - node (v14.15.1, x64)
* Monaco - node (v14.15.1, x64)
* TFS - node (v14.15.1, x64)
* material-ui - node (v14.15.1, x64)
* xstate - node (v14.15.1, x64)
  

**Summary**  
<table cellspacing="0" cellpadding="0" border="0">
<tr>
<th align="left">Benchmark</th>
<th align="left">Name</th>
<th align="left">Iterations</th>
</tr>
<tr>
<td>Current</td>
<td>50811</td>
<td>10</td>
</tr>
<tr>
<td>Baseline</td>
<td>50811</td>
<td>10</td>
</tr>
</table>
</details>

Also of note, I'm going to have to change our bundled output format to an IIFE; typescript uses a single file for both node and the browser, conditionally using `module.exports`. This isn't something that esbuild can do, but it looks like I can just use an IIFE with a footer and tack on some extra code at the bottom to set things up (per https://github.com/evanw/esbuild/issues/507#issuecomment-727690399; TS doesn't need full UMD support, but it's close).

**Comment by jakebailey** [comment]
## 2022-09-29

At this point, everything is mostly done. I ended up writing my own small dts bundler tool that can cover our codebase specifically; we have some unique requirements, namely that we define our own `Node`, `Symbol`, `Map`, etc. types, which most rollup tools try and rename. The limited bundler produces namespaces that look pretty similar to our current d.ts files and makes use of the fact that we don't have any name duplication through our exports, something that is effectively already required by the fact that we do `export *` and that errors if you don't resolve the conflict.

At this point, the only remaining tasks are:

- Backporting changes so the stack isn't so big, including build changes and other removals, such as...
- Changes related to removing `typescriptServices.js` (#50758); this isn't strictly required but simplifies my build a little bit.
- Fixing up `createPlaygroundBuild`, or remove it entirely.
- Get a plan in place to correctly ignore the final commit in `git blame` using `.git-blame-ignore`; this will hopefully make this change hidden from GitHub blames and potentially GitLens or other tools.
- Testing the developer experience when it comes to auto-importing from the new namespace barrel modules; I might have to move them to another location so they are chosen as the "right" place to import helpers from.

**Comment by jakebailey** [comment]
# 2022-10-11

Down to basically odds and ends.

- CodeQL appears to get stuck (https://github.com/microsoft/TypeScript/actions/runs/3229429756/jobs/5286722214); probably something to do with how circular the codebase is. I plan to try and run the same "layer violation" software that the VS Code team used to fixup their codebase, after the fact. It'd be nice to switch to direct, non-cyclical imports.
- I have sent PRs for monaco to fix things up in prep for 5.0: https://github.com/microsoft/monaco-editor/pull/3344, https://github.com/microsoft/monaco-editor/pull/3356, which are already applied in https://github.com/microsoft/TypeScript-Make-Monaco-Builds and running for all playground builds. These changes allow us to remove `typescriptServices.js`, as well as no longer worry about our code changes accidentally breaking the source-level hacks applied in monaco.
  - We have a copy of their import script in our own repo as `createPlaygroundBuild`, but it turns out that monaco's build has changed enough to the point that this script no longer works; monaco needs tsWorker.js, a file that only their build can produce as it includes code from their codebase along with ours.
- I have automated `.git-blame-ignore-revs` to ignore the "conversion step" commits that do the bulk of the work and make all of the noise. When we merge the module branch into main, we'll have to use a merge commit rather than a squash to make this work. Ignoring a squash commit itself will cause git blames to look very odd as there is a lot of new code that I do want git blame to work on.
- Auto-imports appear to work.

We should be good to dogfood this.

Also, since this effectively requires a whole new build, I've taken this opportunity to make a number of major build changes to speed up our build performance, remove dependencies, replace dependencies, and so on; changes that are hard to make during our normal dev cycle without causing churn. The module conversion itself is a huge speedbump to anyone bisecting around as it is.

**Comment by jakebailey** [comment]
## 2022-10-19

Things are looking good, still at odds and ends. The dogfooding has netted some useful feedback, some issues I've fixed like tsserver needing to be a UMD-module for web, and so on.

- The big news this week is that I've been able to shave an additional ***5.7 MB*** off our package (TS 4.8 is 63 MB, post-modules we will be at 35.6 MB, ~43% smaller). This comes from a seemingly unrelated change; we have a `compiler-debug.js` that's build locally which adds in some helpers to format the code flow graph in debugging. This was not included into the compiler as there's no real need to ship this to end users. However, the way that it worked was to require it, then call an `init` function with the `ts` namespace. By moving that code directly into our Debug namespace (~300 lines), we no longer need to pass in the `ts` namespace, which enables bundlers actually tree shake it. Since the Debug namespace is used in _every_ bundle, this means that both `tsc` and `typingsInstaller` shrunk by that 5.7 MB number. Notably, this doesn't affect `typescript.js`, `tsserver.js`, or `tsserverlibrary.js`, as those support the plugin API, which has to pass in the `ts` namespace just like the debug helper did. So, those ones are effectively unable to tree shake, which is unfortunate. But in any case, it's a win.
- I am no longer planning on handling let/const in any way; I considered sending a PR to esbuild that does the "easy" examples, but I believe this is actually a bad idea because in the future we'll likely be emitting ESM and we don't want to emit something like `export var` such that consumers only see mutable bindings.
- All of the changes to monaco have been merged, so our playground will continue to work as expected, and in a way that lets us do things like minification in the future without trouble.
- All of our pipelines are updated in prep for the new task runner.

After some design meeting bikeshedding:

- We're going to target ES2018; I picked this as it was the ES version we list for Node 10, and it doesn't seem like we'd gain much from newer syntax besides nullish colalescing, which only appeared in Node 14. (I'd rather not leave Node 12 people in the dust quite yet.)
  - Related to this, we are also all okay with dropping `ts.Map`, `ts.Set`, etc; with our target no longer being ES5, we can just start using the real types. In a previous version of TS, I deleted our shims entirely, and nobody has been bitten yet, so this seems safe plus/minus the API break.
- We're not going to do anything automated to make imports sorted / have a max line limit. All of the different things we think we could do here are undesirable in various ways; we can't make each import its own line (the files would be huge), we can't use a formatter to reflow them as that would cause major merge conflicts, and so on. This is easy to address in the future, and the reality is that it's rare to add a new import to our worst files like the checker.

Remaining work:

- CodeQL analysis of our repo times out. Still figuring out who to talk to there.
- The debugging experience in VS Code is a little slow; need to talk to the VS Code debugger people about that one.
- My stack needs to be cleaned up again; I've made lots of small changes and since this PR will be merged (not squashed), it should be cleaner.

**Comment by jakebailey** [comment]
## 2022-10-25

Probably the last update before we make the switch in 5.0.

First off, more performance wins; `tsc.js` is now 30% faster to start. Measured using hyperfine, our various entrypoints:

<details><summary>Expand me</summary>

```
Benchmark 1: tsc main
  Time (mean ± σ):     115.3 ms ±   1.1 ms    [User: 103.3 ms, System: 11.5 ms]
  Range (min … max):   112.5 ms … 123.9 ms    1000 runs

Benchmark 2: tsc module-ified
  Time (mean ± σ):      89.2 ms ±   1.1 ms    [User: 78.3 ms, System: 10.1 ms]
  Range (min … max):    86.8 ms … 101.4 ms    1000 runs

Summary
  'tsc module-ified' ran
    1.29 ± 0.02 times faster than 'tsc main'
```

```
Benchmark 1: tsserver main
  Time (mean ± σ):     161.0 ms ±   1.6 ms    [User: 147.1 ms, System: 13.7 ms]
  Range (min … max):   157.4 ms … 170.8 ms    1000 runs

Benchmark 2: tsserver module-ified
  Time (mean ± σ):     151.9 ms ±   1.4 ms    [User: 137.6 ms, System: 15.0 ms]
  Range (min … max):   148.4 ms … 161.9 ms    1000 runs

Summary
  'tsserver module-ified' ran
    1.06 ± 0.01 times faster than 'tsserver main'
```

```
Benchmark 1: typescript main
  Time (mean ± σ):     146.3 ms ±   1.7 ms    [User: 132.4 ms, System: 14.2 ms]
  Range (min … max):   142.9 ms … 165.6 ms    1000 runs

Benchmark 2: typescript module-ified
  Time (mean ± σ):     132.6 ms ±   1.5 ms    [User: 120.0 ms, System: 13.4 ms]
  Range (min … max):   129.6 ms … 147.3 ms    1000 runs

Summary
  'typescript module-ified' ran
    1.10 ± 0.02 times faster than 'typescript main'
```

```
Benchmark 1: tsserverlibrary main
  Time (mean ± σ):     153.0 ms ±   1.3 ms    [User: 139.2 ms, System: 14.1 ms]
  Range (min … max):   149.7 ms … 161.4 ms    1000 runs

Benchmark 2: tsserverlibrary module-ified
  Time (mean ± σ):     143.1 ms ±   1.5 ms    [User: 127.7 ms, System: 16.0 ms]
  Range (min … max):   139.5 ms … 158.4 ms    1000 runs

Summary
  'tsserverlibrary module-ified' ran
    1.07 ± 0.01 times faster than 'tsserverlibrary main'
```

</details>

I have filed bugs for the problems cross-team, linked in the previous post. These are the bugs I've experienced during the development of the conversion:

- https://github.com/microsoft/TypeScript/issues/51301
- https://github.com/microsoft/TypeScript/issues/35004

The former is an annoyance; the latter is a half-blocker towards converting the Debug namespace into a module (which will allow bundlers to directly call our debug helpers).

I also have a big writeup which will go into the final PR's description + be converted into a blog post.

See you all then.

**Comment by Jack-Works** [comment]
![image](https://user-images.githubusercontent.com/5390719/198521256-eadfa0ad-6c1f-45d1-834d-b87b1defd957.png)

still have some require calls. any plan to move it to ts.sys?

**Comment by Jack-Works** [comment]
> I prototyped a Jest runner that could run our compiler tests

please consider jest's snapshot test! manually update some test is pain

**Comment by jakebailey** [comment]
Jack Works Thursday, October 27, 11:47 PM
still have some require calls. any plan to move it to ts.sys?




________________________________

**Comment by Jack-Works** [comment]
> we don't want to emit something like export var such that consumers only see mutable bindings.

they're immutable even if they're let/var anyway

**Comment by jakebailey** [comment]
The PR has been posted! #51387

**Comment by jakebailey** [comment]
The modules PR went in earlier today. Here's a set of follow-up issues and PRs that I am already working on which build on the modules change:

- #51438
- #51439
- #51440
- #51441
- #51442
- #51443
- #51444

**Comment by kungfooman** [comment]
This wiki page is still references `protocol.d.ts": https://github.com/Microsoft/TypeScript/wiki/Standalone-Server-%28tsserver%29

Should it just point to here then? https://github.com/microsoft/TypeScript/blob/main/src/server/protocol.ts

**Comment by jakebailey** [comment]
It should probably just say "check the `ts.server.protocol`" namespace, reflecting the suggestion in my PR. Thanks for finding that.

**Comment by robpalme** [comment]
Hello @jakebailey

The TDZ performance finding quoted here ([the 5% — 9% parser win](https://github.com/microsoft/TypeScript/issues/49332#issuecomment-1251414473)) is being quoted as [one of the reasons to perform TDZ optimisation in ESBuild.](https://github.com/evanw/esbuild/issues/2889#issuecomment-1418240693) Meaning ESBuild is replacing `const` and `let` with `var` to improve performance with no way to opt-out.

We all want faster code and so this optimization may well be justified but I'd like to dig in a bit more before concluding. Because excess downlevelling hampers debugging, as you have previously recognized. And when we abandon using modern features in production via tooling that downlevels, it reduces the incentive for engines to optimize those features - a negative feedback loop.

Specifically I'd like to understand if there is a missing V8 optimisation here. So two questions:

- Have you replicated this performance delta on any other engine?
- Do you have an isolated/minimal repro that could be provided to V8 team as part of a performance bug report?

**Comment by jakebailey** [comment]
> Meaning ESBuild is replacing `const` and `let` with `var` to improve performance with no way to opt-out.

Having read that linked thread, I honestly don't know how esbuild could _stop_ using `var` without some major performance changes in v8 and such; the `var` + lazy init function pattern is explicitly one where if `let` were used, the TDZ checks would affect basically use.

> because excess downlevelling hampers debugging, as you have previously recognized.

This is true, but in this particular case, switching to var does not really hurt debugging so long as the vars that are switched don't generate any other additional code such as the extra closures needed for `for` loop block scoping. But, it did catch at least one bug.

> Have you replicated this performance delta on any other engine?

The benchmark suite that tests TypeScript largely runs tsc, and tsc only runs in node, so I have not tested anything other than v8, no. Perhaps recent Bun changes mean I can do a drop-in test and see what happens. (I'm pretty sure Deno is v8 so I wouldn't test that.)

> Do you have an isolated/minimal repro that could be provided to V8 team as part of a performance bug report?

I don't; what I really need is some sort of tool that acts as a "TDZ blame" and can tell me where all of my TDZ time is being spent, but no such tool exists. The parser is very large and besides simply running our code through a transform which changes all `const`/`let` to `var`, it's hard for me to know what I can do to actually find the problem than to sit down and just replace a bunch and see what happens.

**Comment by jakebailey** [comment]
> it's hard for me to know what I can do to actually find the problem than to sit down and just replace a bunch and see what happens.

That being said, I could likely do a differential profile between the two and look there.

**Comment by jakebailey** [comment]
@robpalme Just to show the current state, https://github.com/microsoft/TypeScript/pull/52656#issuecomment-1421249664 is what it'd be like if we downleveled all let/const to var; the performance difference is really stunning.

I still need to go try this on another engine, but the delta shows that there really is a lot of perf sitting on the table.
